janusgraph-test/src/test/java/org/janusgraph/graphdb/query/QueryTest.java

```diff
@@ -86,7 +86,7 @@ public void testQueryLimitAdjustment() {
 
         for (int i = 0; i < 20; i++) {
             tx.addVertex().property("prop1", "prop1val").element().property("prop2", "prop2val")
-                .element().property("prop3", "prop3val" + i % 2).element().property("prop4", "prop4val" + i % 2);
+                .element().property("prop3", "prop3val").element().property("prop4", "prop4val");
```

Comment:
This change is necessary otherwise tests fail occasionally. The reason is, the in-memory backend (seemingly) does not return results in a consistent order. Therefore, for a query that has two conditions and requires in-memory filtering, the query sometimes needs to progressively fetch more results to satisfy the number of results required. As a matter of consequence, profiling results might not necessarily be consistent.

This was not a problem (never failed before) because the profiling results were incomplete (only covered the first backend query).

Reply:
1. I gave a wrong statement about the inconsistent order. After some experiments, it seems that the in-memory backend indeed returns in a consistent order. What actually happens is, the order of results is not consistent with the order of insertions. To be concrete, the following test would fail:

```
        JanusGraphManagement mgmt = graph.openManagement();
        PropertyKey prop1Key = mgmt.makePropertyKey("prop1").dataType(String.class).make();
        mgmt.buildIndex("prop1_idx", Vertex.class).addKey(prop1Key).buildCompositeIndex();
        mgmt.commit();

        List<Object> vertexIds = new ArrayList<>();
        for (int i = 0; i < 100; i++) {
            Vertex v = tx.addVertex();
            v.property("prop1", "val");
            vertexIds.add(v.id());
        }
        tx.commit();

        List<Object> actualIds = graph.traversal().V().has("prop1", "val").id().toList();
        assertEquals(vertexIds, actualIds);
```

This is (at least partially) because, when JanusGraph commits, it uses an unordered HashSet to store all the mutations. Thus, in the query `graph.traversal().V().has("prop1", "prop1val").has("prop3", "prop3val0").limit(5)`, there is no guarantee whether first 10 results of `V().has("prop1", "prop1val")` has at least 5 results that satisfy `has("prop3", "prop3val0")` condition or not.

2. To my best memory, I didn't use "i % 2" for any particular use. I simply didn't want to use a constant value, to make the vertices look slightly more normal.

Reply:
Thanks for working that out in such a detailed way!